core/src/main/java/org/jivesoftware/sparkimpl/certificates/SparkTrustManager.java

```diff
@@ -0,0 +1,387 @@
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.jivesoftware.spark.util.log.Log;
+import org.jivesoftware.sparkimpl.settings.local.LocalPreferences;
+import org.jivesoftware.sparkimpl.settings.local.SettingsManager;
+
+/**
+ * This trust manager wrap around SparkExceptionsTrustManager. In case when SparkExceptionsTrustManager fail then this
+ * TrustManager will provide certificates from TRUSTED KeyStore which are checked against data validity, revocation,
+ * acceptance of self-signed certificates and basic constraints.
+ * 
+ * 
+ * @author A
+ *
+ */
+public class SparkTrustManager implements X509TrustManager {
+
+    private LocalPreferences localPref = SettingsManager.getLocalPreferences();
+    private boolean checkCRL;
+    private boolean checkOCSP;
+    private boolean acceptExpired;
+    private boolean acceptNotValidYet;
+    private boolean acceptRevoked;
+    private boolean acceptSelfSigned;
+
+    private CertStore crlStore;
+    private Collection<X509CRL> crlCollection = new ArrayList<>();
+    private X509TrustManager exceptionsTrustManager;
+    private Provider bcProvider = new BouncyCastleProvider(); // bc provider for path validation
+    private KeyStore trustStore;
+    private CertificateController certControll = new CertificateController(localPref);
+
+    public SparkTrustManager() {
+        exceptionsTrustManager = new SparkExceptionsTrustManager();
+        Security.addProvider(bcProvider);
+
+        checkCRL = localPref.isCheckCRL();
+        checkOCSP = localPref.isCheckOCSP();
+        acceptExpired = localPref.isAcceptExpired();
+        // acceptNotValidYet = localPref.isAcceptNotValidYet();
+        acceptRevoked = localPref.isAcceptRevoked();
+        acceptSelfSigned = localPref.isAcceptSelfSigned();
+
+        loadTrustStore();
+    }
+
+    public static X509TrustManager[] getTrustManagerList() {
+        return new X509TrustManager[] { new SparkTrustManager() };
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        // TODO Auto-generated method stub
+
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        try {
+            // first check if certificate is accepted as as certificate from exceptions list, exceptionsTrustManager
+            // will make use of chain provided by exceptions KeyStore
+            exceptionsTrustManager.checkServerTrusted(chain, authType);
+        } catch (CertificateException ex) {
+            // in case when certificate isn't on exceptions list then make use of this Trust Manager
+
+            // validate chain by date (expired/not valid yet)
+            checkDateValidity(chain);
+
+            // check if certificate isn't self signed, self signed certificate still have to be in TrustStore to be
+            // accepted
+            if (chain.length > 1) {
```

Comment:
Chain length is not a safe way to determine if the certificate in the chain is self-signed. The chain could be incomplete (it might depend on certificates that are in the truststore, without including those certificates), and chains might contain certificates that are not part of the chain at all (and are unused). You should find a better way to identify "self-signed"